cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/RemoteArtifactManager.java

```diff
@@ -63,14 +66,23 @@
   private final NamespaceId namespaceId;
   private final RetryStrategy retryStrategy;
   private final RemoteClient remoteClientInternal;
+  private final ArtifactLocalizerClient artifactLocalizerClient;
 
-  @Inject
+  @AssistedInject
   RemoteArtifactManager(CConfiguration cConf, LocationFactory locationFactory, @Assisted NamespaceId namespaceId,
                         @Assisted RetryStrategy retryStrategy, RemoteClientFactory remoteClientFactory) {
+    this(cConf, locationFactory, namespaceId, retryStrategy, remoteClientFactory, null);
+  }
+
+  @AssistedInject
+  RemoteArtifactManager(CConfiguration cConf, LocationFactory locationFactory, @Assisted NamespaceId namespaceId,
+                        @Assisted RetryStrategy retryStrategy, RemoteClientFactory remoteClientFactory,
+                        @Assisted @Nullable ArtifactLocalizerClient artifactLocalizerClient) {
```

Comment:
I am wondering why do you use `@Assisted` for `ArtifactLocalizerClient`. If it's optional you can use Optional<ArtifactLocalizerClient> and guice should provide it correctly using `com.google.inject.multibindings.OptionalBinder`. Or you can ask for client to be present every time, allow `getArtifactLocation` to return null and provide a Noop client

Reply:
Since we are already having @Assisted parameters, it was easier to provide the artifactLocalizerClient while creating the RemoteArtifactManager. The class calling ArtifactManagerFactory.create() will exactly know if ArtifactLocalizerClient is required or not. Is there a downside to using @Assisted over Optional?

Reply:
It's mostly that the factory caller should not care to pass the client. Now you need to introduce another factory method and ask caller "take this client from injector and give to me". It should not be caller business, should it?

Reply:
Switched to OptionalBinder